File: images/metadata-backup/backup.py
#!/usr/bin/env python3
#
# Copyright 2023 Flant JSC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from datetime import datetime

import argparse
import base64
import kubernetes
import os
import re
import tarfile
import tempfile
import yaml

objGroup = 'storage.deckhouse.io'
objKind = 'ReplicatedStorageMetadataBackup'
objKindPlural = 'replicatedstoragemetadatabackups'
objVersion = 'v1alpha1'
linstorCRDGroup = 'internal.linstor.linbit.com'
srvBackupLabelKey = 'sds-replicated-volume.deckhouse.io/sds-replicated-volume-db-backup'
backup_type = os.getenv('BACKUP_TYPE')

def create_backup(backup_date, labels={}):
    temp_path = tempfile.mkdtemp()
    backup_archive = tarfile.open(f'{temp_path}/sds-replicated-volume-{backup_type}-{backup_date}.tar', "w")

    for item in kubernetes.client.ApiextensionsV1Api().list_custom_resource_definition().items:
        if item.spec.group != linstorCRDGroup:
            continue

        crd_name_plural = item.spec.names.plural
        version = item.spec.versions[-1]
        crd_version = version.name

        custom_object = kubernetes.client.CustomObjectsApi().list_cluster_custom_object(group=linstorCRDGroup,
                                                                                        plural=crd_name_plural,
                                                                                        version=crd_version)

        with open(f'{temp_path}/{crd_name_plural}.yaml', 'w') as file:
            yaml.dump(custom_object, file)
        backup_archive.add(f'{temp_path}/{crd_name_plural}.yaml')

    backup_archive.close()

    with open(f'{temp_path}/sds-replicated-volume-{backup_type}-{backup_date}.tar', 'rb') as backup_archive:
        encoded_string = base64.b64encode(backup_archive.read())

    step = 262144 # 256KB
    for chunk_pos in range(0, len(encoded_string) // step + 1, 1):
        objName = f'sds-replicated-volume-{backup_type}-{backup_date}-backup-{chunk_pos}'
        isCreated = kubernetes.client.CustomObjectsApi().create_cluster_custom_object(group=objGroup,
                                           version=objVersion,
                                           plural=objKindPlural,
                                           body={
                                             'apiVersion': f'{objGroup}/{objVersion}',
                                             'kind': objKind,
                                             'metadata': {'name': objName},
                                             'spec': {'data': encoded_string[chunk_pos * step:(chunk_pos + 1) * step].decode("utf-8")}}
                                           )
        if not isCreated:
            print(f"backup_job: sds-replicated-volume-{backup_type}-{backup_date}-backup-{chunk_pos} creation failure")
            return False


    current_backup_objects = kubernetes.client.CustomObjectsApi().list_cluster_custom_object(group=objGroup,
                                                                                     version=objVersion,
                                                                                     plural=objKindPlural)
    regex = re.compile(f'sds-replicated-volume-{backup_type}-{backup_date}-backup-')
    for item in current_backup_objects['items']:
        if regex.match(item.get('metadata', {}).get('name', '')):
            kubernetes.client.CustomObjectsApi().patch_cluster_custom_object(
                group=objGroup,
                version=objVersion,
                plural=objKindPlural,
                name=item['metadata']['name'],
                body={"metadata": {"labels": labels}})

    for root, dirs, files in os.walk(temp_path, topdown=False):
        for name in files:
            os.remove(os.path.join(root, name))
        for name in dirs:
            os.rmdir(os.path.join(root, name))

    return True


if __name__ == "__main__":
    full_completed_list = []
    backup_list = []

    parser = argparse.ArgumentParser("backup")
    parser.add_argument("-r", "--retentionCount", help="Retention days count", type=int)
    args = parser.parse_args()
    retention = args.retentionCount or 7

    kubernetes.config.load_incluster_config()

    date_time = datetime.now().strftime("%Y%m%d%H%M%S")
    success = create_backup(backup_date=date_time, labels={srvBackupLabelKey: f'{date_time}', "completed": "true"})
    if not success:
        print(f"backup_job: creation failure")
        exit(1)

    current_backup_objects = kubernetes.client.CustomObjectsApi().list_cluster_custom_object(group=objGroup,
                                                                                     version=objVersion,
                                                                                     plural=objKindPlural)
    for item in current_backup_objects['items']:
        backup_list.append(item['metadata']['name'])
        if re.search(f'^sds-replicated-volume-{backup_type}-\\d+-backup-0$', item['metadata']['name']):
            full_completed_list.append(item['metadata']['name'])
    completed_list = full_completed_list[-retention:]
    for backup_name in backup_list:
        if not re.search(f'^sds-replicated-volume-{backup_type}-\\d+-backup-', backup_name):
            continue
        backup_cut_name = re.search(f'^sds-replicated-volume-{backup_type}-\\d+-backup-', backup_name).group(0)
        if f'{backup_cut_name}0' in completed_list:
             continue
        kubernetes.client.CustomObjectsApi().delete_cluster_custom_object(group=objGroup,
                                                                          version=objVersion,
                                                                          plural=objKindPlural,
                                                                          name=backup_name)

File: images/linstor-migrator/internal/linstorbackup/backup.go
/*
Copyright 2025 Flant JSC

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Package linstorbackup dumps LINSTOR API CustomResourceDefinitions and their custom resources
// (internal.linstor.linbit.com) to gzipped multi-document YAML.
package linstorbackup

import (
	"bytes"
	"compress/gzip"
	"context"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"sort"

	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
	apierrors "k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/runtime"
	"k8s.io/apimachinery/pkg/runtime/schema"
	"k8s.io/client-go/dynamic"
	"sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/yaml"

	"github.com/deckhouse/sds-replicated-volume/images/linstor-migrator/internal/config"
)

const (
	// APIGroup is the LINSTOR Kubernetes API group backed up from CRD discovery (spec.group).
	APIGroup = "internal.linstor.linbit.com"

	backupCRDsGzFile = "crds.gz"
	backupCRsGzFile  = "crs.gz"
	backupReadme     = "readme.txt"
)

func backupReadmeText() string {
	exampleDir := filepath.Join(config.MigratorHostDir, config.MigratorLinstorBackupDirName)
	exampleCRDs := filepath.Join(exampleDir, backupCRDsGzFile)
	exampleCRs := filepath.Join(exampleDir, backupCRsGzFile)
	return fmt.Sprintf(`LINSTOR backup (API group internal.linstor.linbit.com)
========================================================

Files:
  crds.gz — gzip multi-document YAML: one CustomResourceDefinition per document
            (spec.group internal.linstor.linbit.com).
  crs.gz  — gzip multi-document YAML: every custom resource instance for those CRDs.

Restore:
  1. mkdir -p /tmp/linstor-restore && cd /tmp/linstor-restore
  2. gunzip -c %s > crds.yaml
  3. kubectl apply -f crds.yaml
  4. gunzip -c %s > crs.yaml
  5. kubectl apply -f crs.yaml

Notes:
  - Apply CRDs before CRs. Status and server-owned metadata (managedFields, resourceVersion, uid,
    creationTimestamp, generation) are omitted from both files so manifests are apply-friendly.
`, exampleCRDs, exampleCRs)
}

// BackupDir returns the host path for LINSTOR CR backups under MigratorHostDir.
func BackupDir() string {
	return filepath.Join(config.MigratorHostDir, config.MigratorLinstorBackupDirName)
}

// WriteGroupBackup lists all CustomResourceDefinitions for APIGroup, lists every instance of each
// resource, and writes crds.gz (CRD definitions), crs.gz (custom resources), and readme.txt into dir.
func WriteGroupBackup(ctx context.Context, kClient client.Client, dyn dynamic.Interface, log *slog.Logger, dir string) error {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("create backup directory %q: %w", dir, err)
	}

	crds := &apiextensionsv1.CustomResourceDefinitionList{}
	if err := kClient.List(ctx, crds); err != nil {
		return fmt.Errorf("list CustomResourceDefinitions: %w", err)
	}

	var linstorCRDs []apiextensionsv1.CustomResourceDefinition
	for i := range crds.Items {
		crd := crds.Items[i]
		if crd.Spec.Group != APIGroup {
			continue
		}
		linstorCRDs = append(linstorCRDs, crd)
	}
	sort.Slice(linstorCRDs, func(i, j int) bool { return linstorCRDs[i].Name < linstorCRDs[j].Name })

	// Serialize each CustomResourceDefinition (apiextensions) into the multi-doc YAML for crds.gz.
	var crdYamlBuf bytes.Buffer
	for i := range linstorCRDs {
		crd := &linstorCRDs[i]
		b, err := crdToApplyFriendlyYAML(crd)
		if err != nil {
			return fmt.Errorf("marshal CRD %q: %w", crd.Name, err)
		}
		if i > 0 {
			crdYamlBuf.WriteString("---\n")
		}
		crdYamlBuf.Write(b)
		log.Debug("backed up LINSTOR CRD definition", "crd", crd.Name)
	}
	crdGzPath := filepath.Join(dir, backupCRDsGzFile)
	if err := writeGzipFile(crdGzPath, crdYamlBuf.Bytes()); err != nil {
		return err
	}

	// List and serialize every custom resource instance per LINSTOR CRD into the multi-doc YAML for crs.gz.
	var crsYamlBuf bytes.Buffer
	crDocCount := 0
	for _, crd := range linstorCRDs {
		ver, err := storageVersionName(&crd)
		if err != nil {
			return fmt.Errorf("CRD %q: %w", crd.Name, err)
		}
		gvr := schema.GroupVersionResource{
			Group:    crd.Spec.Group,
			Version:  ver,
			Resource: crd.Spec.Names.Plural,
		}
		res := dyn.Resource(gvr)
		var list *unstructured.UnstructuredList
		if crd.Spec.Scope == apiextensionsv1.ClusterScoped {
			list, err = res.List(ctx, metav1.ListOptions{})
		} else {
			list, err = res.Namespace(metav1.NamespaceAll).List(ctx, metav1.ListOptions{})
		}
		if err != nil {
			if apierrors.IsForbidden(err) {
				return fmt.Errorf("list %s/%s: %w", gvr.GroupVersion().String(), gvr.Resource, err)
			}
			return fmt.Errorf("list %s/%s: %w", gvr.GroupVersion().String(), gvr.Resource, err)
		}
		for _, item := range list.Items {
			clean := item.DeepCopy()
			stripApplyUnfriendlyFields(clean.Object)
			b, err := yaml.Marshal(clean.Object)
			if err != nil {
				return fmt.Errorf("marshal %s %s/%s: %w", crd.Name, clean.GetNamespace(), clean.GetName(), err)
			}
			if crDocCount > 0 {
				crsYamlBuf.WriteString("---\n")
			}
			crsYamlBuf.Write(b)
			crDocCount++
		}
		log.Debug("backed up LINSTOR custom resources", "crd", crd.Name, "count", len(list.Items))
	}
	crsGzPath := filepath.Join(dir, backupCRsGzFile)
	if err := writeGzipFile(crsGzPath, crsYamlBuf.Bytes()); err != nil {
		return err
	}

	readmePath := filepath.Join(dir, backupReadme)
	if err := os.WriteFile(readmePath, []byte(backupReadmeText()), 0o644); err != nil {
		return fmt.Errorf("write readme: %w", err)
	}

	log.Info("wrote LINSTOR backup",
		"dir", dir,
		"crdFile", backupCRDsGzFile, "crdDocuments", len(linstorCRDs),
		"crFile", backupCRsGzFile, "crDocuments", crDocCount,
	)
	return nil
}

func crdToApplyFriendlyYAML(crd *apiextensionsv1.CustomResourceDefinition) ([]byte, error) {
	u, err := runtime.DefaultUnstructuredConverter.ToUnstructured(crd.DeepCopy())
	if err != nil {
		return nil, err
	}
	stripApplyUnfriendlyFields(u)
	// controller-runtime typed List/Get often leaves embedded TypeMeta empty, so ToUnstructured
	// omits apiVersion/kind; kubectl apply requires them.
	u["apiVersion"] = apiextensionsv1.SchemeGroupVersion.String()
	u["kind"] = "CustomResourceDefinition"
	return yaml.Marshal(u)
}

// stripApplyUnfriendlyFields removes status and server-populated metadata so kubectl apply restore works.
func stripApplyUnfriendlyFields(obj map[string]interface{}) {
	unstructured.RemoveNestedField(obj, "metadata", "managedFields")
	unstructured.RemoveNestedField(obj, "metadata", "resourceVersion")
	unstructured.RemoveNestedField(obj, "metadata", "uid")
	unstructured.RemoveNestedField(obj, "metadata", "creationTimestamp")
	unstructured.RemoveNestedField(obj, "metadata", "generation")
	unstructured.RemoveNestedField(obj, "status")
}

func writeGzipFile(path string, data []byte) error {
	f, err := os.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o644)
	if err != nil {
		return fmt.Errorf("create %q: %w", path, err)
	}
	gzw := gzip.NewWriter(f)
	if _, err := gzw.Write(data); err != nil {
		_ = gzw.Close()
		_ = f.Close()
		return fmt.Errorf("gzip write %q: %w", path, err)
	}
	if err := gzw.Close(); err != nil {
		_ = f.Close()
		return fmt.Errorf("gzip close %q: %w", path, err)
	}
	if err := f.Close(); err != nil {
		return fmt.Errorf("close %q: %w", path, err)
	}
	return nil
}

func storageVersionName(crd *apiextensionsv1.CustomResourceDefinition) (string, error) {
	for _, v := range crd.Spec.Versions {
		if v.Storage {
			return v.Name, nil
		}
	}
	if len(crd.Spec.Versions) > 0 {
		return crd.Spec.Versions[0].Name, nil
	}
	return "", fmt.Errorf("no versions")
}
